docs/proposals/memory-management.md

```diff
@@ -0,0 +1,135 @@
+
+Cgroups are used to place memory limits to containers. It is probably reasonable to expect that most users don't set memory limits. Docker is used to track OOM kills in containers. Docker does not cleanup the entire container when an OOM kill is encountered. If the pid `1` in the container gets killed, the container is cleaned up immediately. If not, the container has the opportunity to handle the OOM kill gracefully (e.x. stress container). Kubelet currently surfaces the OOM kill activity to users via Kubernetes APIs. 
+
+Quality of Service was recently introduced. There are three classes of service:
+
+1. Guaranteed - Memory requests are guaranteed
+
+2. Burstable - Memory limits guaranteed up to user request
+
+3. Best Effort - No guaranty - memory availability can change over time.
+
+The scheduler does not take into account the total usage on the node. 
+Swapping is turned off by default. Kubelet does not track or manage swap files or partitions.
+Kubelet (via cadvisor) can track the memory usage of system daemons and pods. It does not take any action based on usage. 
+
+The kernel OOM killer is expected to be enabled. oom_score_adj is used to protect guaranteed jobs and system daemons. Burstable jobs are given a oom_score_adj that will result in them having a oom_score of `1000` as soon as they hit their original request, making them a likely candidate for being killed by the oom killer. Best effort tasks always have a `oom_score_adj` of `1000` which makes them the most likely candidates to be killed. The kernel usually chooses the highest memory consuming process ([along with a few other policies](https://lwn.net/Articles/317814/))
+
+####Issues with Current model
+The life of a node is blissful up until it has memory pressure, at which points the kernel resorts to killing processes.
+
+* Memory is inherently overcommitted.
+
+ This combined with lack of swap space, increases the probability of system OOMs. This slows down the responsiveness of the system and could have secondary effects like high tail latencies for system daemons.
+
+* A container in a pod with Restart Always policy will be restarted continuously even if it OOMs. 
+
+* Other than the Guaranteed class, jobs in other classes don't have a memory limit.
+
+* Any process in a container can be killed due to either system memory pressure or exceeding container limits.
+
+ There is no container cleanup policy in the event of an OOM kill. What should happen to a pod if a process in one of its container's was OOM killed? 
+
+* Kubelet cannot associate a system memory pressure OOM kill to a specific container or Pod. 
```

Comment:
There are two kinds of OOM kills.
1. Whenever a cgroup is about to exceed its memory limit, a process from that cgroup will be killed. Kubelet has the ability to detect this.
2. Whenever the overall system memory goes very low, the kernel will evict some process from the system. This process can be in any container. Kubelet cannot detect this.

I will add this to the doc. 

@bgrant0607: I will add a link to the API as well